### ISSUE TITLE
Single file `saveAs*` functions

### ISSUE THREAD

**Issue by martinbomio** [spotify/scio#2752]
We have some use cases where we want to save only one object to a file, usually with a fixed name, removing the sharding information stored in the file name. We do it for JSON file or a tf.Schema file. 

We end up writing implicit classes that extend the scio context (for reading) and Scollection of our type (for writing) and we thought it might be something that other people might want.

We could also implement those with beam's `FileSystem`, but this will mean we cannot test it with `JobTest`

**Comment by regadas** [comment]
Hi @martinbomio, can you provide an example of what you're doing?

Also have you consider using `saveAsCustomOutput ` instead of implicit class `SCollection`?

**Comment by martinbomio** [comment]
Hi @regadas these are two examples I've ended up doing in the past:
```scala
    def saveAsSchemaSnapshotProtobuf(
      path: String,
      name: String = "schema_snapshot.pb",
      compression: Compression = Compression.UNCOMPRESSED
    ): Unit =
      if (col.context.isTest) {
        col.map(_.toByteArray).saveAsBinaryFile(path + "/" + name, compression = compression)
      } else {
        val io = FileIO
          .write[Array[Byte]]
          .to(path)
          .via(new BytesSink())
          .withCompression(compression)
          .withNumShards(1)
          .withNaming(fixedFileName(name))
        col.map(_.toByteArray).internal.apply(io)
      }
```
and 
```
    def saveAsSingleTextFile(
      path: String,
      header: Option[String] = None,
      suffix: String = ".tsv",
      compression: Compression = Compression.UNCOMPRESSED
    ): Unit = {
      if (self.context.isTest) {
        self.saveAsTextFile(path, suffix = suffix, compression = compression)
      } else {
        val dest = header match {
          case Some(h) =>
            TextIO
              .write()
              .to(path)
              .withoutSharding()
              .withSuffix(suffix)
              .withHeader(h)
              .withWritableByteChannelFactory(
                FileBasedSink.CompressionType.fromCanonical(compression)
              )
          case _ =>
            TextIO
              .write()
              .to(path)
              .withoutSharding()
              .withSuffix(suffix)
              .withWritableByteChannelFactory(
                FileBasedSink.CompressionType.fromCanonical(compression)
              )
        }

        self.internal.apply(dest)
      }
```

I didn't know about `saveAsCustomOutput`, it will be nice to use it inside the implicit class, still think that having the implicit class which will contain the `PTransform` is nice.

**Comment by jto** [comment]
You can also implement your own IO and use it with `SCollection.write` See [`CsvIO`](https://github.com/spotify/scio/blob/master/scio-extra/src/main/scala/com/spotify/scio/extra/csv/CsvIO.scala) for an example.

**Comment by martinbomio** [comment]
Yeah, there are multiple ways this can be implemented. The one we have right now is working for us. I was more curious to see if other people also end up writing their own thing, in which case, it might make sense to be provided by `scio`

**Comment by regadas** [comment]
Hi @martinbomio!  it seems that we a small minor changes to `saveAsTextFile` and `saveAsBinaryFile` we can accommodate your use case better. Let me get back to you on this ...

**Comment by regadas** [comment]
@martinbomio #2758 #2759

**Comment by regadas** [comment]
Closed via #2758 #2759
